enhancements/storage/csi-driver-operator-merge.md

```diff
@@ -0,0 +1,530 @@
+  do it ourselves.
+
+Advantages:
+
+* Keeping image names, i.e. no new Comet repos (distgits, brew packages, etc.)
+* Harder (but not impossible) to test before the switch. In ideal case, we're switching just the source repository
+  of the same image, the actual operator code should be the same.
+
+#### Gradual merge
+
+We will merge the operators one by one, starting with AWS EBS. Second will be Azure Disk and Azure File, as we need to
+support HyperShift for them and we want to share code with AWS EBS. The rest will follow as time allows, possibly over
+multiple OCP releases.
+
+Very provisional and optimistic plan:
+
+* AWS EBS, Azure Disk and Azure File in 4.15.
+* The rest in 4.16 or later.
+
+#### Generated YAML files
+
+All CSI driver YAML files look very similar today. We plan to generate them from a single set of templates, so that we
+don't need to maintain them separately. Exact details about the generator are TBD. We want something like kustomize,
+but better integrated with the operator code.
+
+### Risks and Mitigations
+
+* Since each CSI driver operator is different, it may be hard to merge all their code. Some differences are:
+  * Shared Resource (and maybe others) deploy an extra validating webhook.
+  * Cinder (and maybe others) syncs Secrets to a different format.
+  * Manila installs the CSI driver only optionally.
+  * AWS sync CA-bundle from a different namespace.
+  * vSphere interacts with vSphere API a lot.
+  * AWS EFS and GCP Filestore are OLM operators.
+  * And many others.
+
+  All the operators use shared code from library-go already, so merging them to a single repository will not make them
+  worse. Still, it may be more difficult to share even more code e.g. for HyperShift.
+
+* Since we plan to generate also YAML files for the CSI drivers, there is a risk that an exotic CSI driver will
+  require a different YAML file than the others. We will keep possibility for a CSI driver to provide its own YAML file,
+  not using the generator.
+
+### Drawbacks
+
+* We will need to support old images + github repos in all supported z-streams for quite some time, so the real benefit
+  will be visible only after few years. Until then, we will have to maintain repositories with the old CSI driver
+  operators _and_ the new merged repository.
+
+* While each CSI driver operator will be a separate binary and a separate image, all their dependencies will be in a
+  single repository. `vendor/` dir of this repository will be large, as it will contain many (all?) SDKs of clouds that
+  we support.
+  * There is a risk that the operators will require different versions of a vendored package. So far, we kept library-go
+    and k8s packages at the same version in all CSI driver operators without issues, but we did not monitor _all_ the
+    packages.
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
```

Comment:
Should we make it a goal that we can test operators individually? i.e. a change in an AWS specific file does not trigger CI for GCP for example.

Reply:
Good point, added.